### ISSUE TITLE
Failing to start Ceph monitor service on CentOS

### ISSUE THREAD

**Issue by rnowling** [ceph/ceph-ansible#790]
Using `bentos/centos-7.1` with Infernalis or Jewel, the Ceph monitor service is failing to start:

```
fatal: [mon2]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_args": {"arguments": "", "enabled": true, "name": "ceph-mon@ceph-mon2", "pattern": null, "runlevel": "default", "sleep": null, "state": "started"}, "module_name": "service"}, "msg": "Error when trying to enable ceph-mon@ceph-mon2: rc=1 Failed to issue method call: No such file or directory"}
```

**Comment by leseb** [comment]
Could this be because of https://github.com/ceph/ceph-ansible/pull/781?
Are you using vagrant?

**Comment by rnowling** [comment]
Yes, I'm using vagrant.  None of the changes in #781 touch the `ceph-mon` service if I understand correctly.  I'll pull the latest version with that commit and give it a try.

**Comment by leseb** [comment]
argh correct

**Comment by rnowling** [comment]
This still fails with the latest commit (746a38).  I can get it to work if I move `/usr/lib/systemd/system/ceph-mon@.service` to `/usr/lib/systemd/system/ceph-mon.service` and updated `roles/ceph-mon/tasks/start_monitor.yml` to use `ceph-mon` instead of `ceph-mon@{{ monitor_name }}` .  This suggests an issue with the systemd version in the VM image and may need to be moved to upstream Ceph.

**Comment by ktdreyer** [comment]
"`Failed to issue method call: No such file or directory`" - Where is that error coming from? Ansible or Ceph?

Moving `/usr/lib/systemd/system/ceph-mon@.service` to `/usr/lib/systemd/system/ceph-mon.service` doesn't sound right...

**Comment by rnowling** [comment]
@ktdreyer Not claiming that is the right solution. :) Just an observation that may be useful in debugging the issue.

I modified the service module to print more details to get the full command:

```
/bin/systemctl enable ceph-mon@ceph-mon0
```

When I run that from an SSH session, I get the same error:

```
$ sudo /bin/systemctl enable ceph-mon@ceph-mon2
Failed to issue method call: No such file or directory
```

I confirmed that I could run `sudo /bin/systemctl` to make sure it wasn't an issue with the `systemctl` binary being on the path.  (Not likely an issue but just in case)

So, after some further digging, I realized that `systemctl` gives a different error if a unit isn't found and the variable `%i` is used in in the init files for the text after `@`.  As such, it's possible the error is coming from something in the init file itself.

**Comment by rnowling** [comment]
Following the example, in the `start_docker_monitor.yml`, I created a link `/etc/systemd/system/ceph-mon@ceph-mon0.service` to `/usr/lib/systemd/system/ceph-mon@.service` which also solves the problem.   However, I'm not sure if this is a proper solution or not and why this is required for me but doesn't seem to be needed by others.

**Comment by ktdreyer** [comment]
`systemctl enable ceph-mon@ceph-mon0` should be creating that symlink for you automatically (and then reloading systemd so that it takes effect).

**Comment by rnowling** [comment]
A similar issue ( #642 ) was filed previously.  The PR ( #658 ) added a symlink.  So even if `systemctl enable` is supposed to do that, it doesn't seem to be doing that.

**Comment by andrewschoen** [comment]
@rnowling have you tried with any other vagrant boxes? I just tried ``vagrant up`` with ``boxcutter/centos72`` on master and the playbook completed successfully.

**Comment by rnowling** [comment]
@andrewschoen I'm using bento . The official CentOS boxes failed on me with the virtual box provisioner due to SATA Controller (tried with abs without the space) missing. I'll give the box cutter and puppet labs images a try. Would be nice just to use official CentOS images

**Comment by rnowling** [comment]
I was able to successfully provision a Jewel cluster from master using the `boxcutter/centos72` image.  I had been using the `bento/centos-7.1` image specified in the comments of `vagrant_variables.yml.sample`.  Just to sanity check, I created a new set of VMs using `bento/centos-7.1`, ran `sudo yum update -y` on each VM, and attempted to provision -- it worked!  Thus, I think the issue is that the CentOS base images were too old.  Since CentOS runs on a rolling release, however, any version of CentOS will be updated to the most recent with an update.

So, a few thoughts:
1) Update the documentation to mention that the latest version should be used.  Hard to keep specific versions up to date in docs so maybe docs should exclude specific version.

2) The official CentOS images are kept up to date, so `centos/7` should always be the latest.   Recommending the official centos images is properly a good idea, anyway.  I see that the `centos` images don't work with virtualbox and may require some changes.  kbsingh and other CentOS community members are very responsive and have updated boxes to support larger sparse disks and such for me in the past.  I imagine he would be happy to work with us to figure out what is needed to get the official images working under virtualbox.

3) Add tasks for updating the images before provisioning.  I'm not sure how this will affect users of other images (Atomic, Ubuntu, etc.), but we might be able to find an acceptable solution here.

What do you guys think?

**Comment by ktdreyer** [comment]
> 2) The official CentOS images are kept up to date, so centos/7 should always be the latest.
> Recommending the official centos images is properly a good idea, anyway.

:+1:  to using the official, up-to-date CentOS images.

**Comment by rnowling** [comment]
I see that there is a config option for updating CentOS packages... so maybe just need some docs around that or make it a default if the OS is CentOS.

**Comment by ibravo** [comment]
Run with commit 84dad9a after performing yum update on the physical box, and still failing here:

TASK [ceph-mon : collect admin and bootstrap keys] *****************************
task path: /home/ceph-deploy/ceph-ansible/roles/ceph-mon/tasks/start_monitor.yml:48

My boxes:
CentOS Linux release 7.2.1511 (Core) 
Linux ceph03bl09.hq.ltg 3.10.0-229.el7.x86_64 #1 SMP Fri Mar 6 11:36:42 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

**Comment by leseb** [comment]
@ibravo unless your're using RHCS you shouldn't be impacted.
Moreover the commit is a fix so I don't see how it can impact your deployment.

**Comment by FranBarrera** [comment]
Stuck on [ceph-mon : collect admin and bootstrap keys]
Instance Ubuntu 16.04 on Openstack

**Comment by leseb** [comment]
@FranBarrera I just bootstrapped an Ubuntu VM and I can't reproduce your issue :/

```
Distributor ID: Ubuntu
Description:    Ubuntu 16.04 LTS
Release:        16.04
Codename:       xenial
```

**Comment by FranBarrera** [comment]
@leseb I'm testing with two Ubuntu server 16.04 LTS, one for monitor and other for osd and it's stuck on collect admin and bootsrap keys.

**Comment by leseb** [comment]
@rnowling that's weird, did you investigate a bit?
The monitor clearly did not start, can you check why?
systemd issue? config issue?

Thanks!

**Comment by rootfs** [comment]
@rnowling do you still have this issue? I tested successfully on centos with ceph container in #812

**Comment by rnowling** [comment]
@leseb The main thing I noticed is that systemd didn't seem to recognize the `ceph-mon@X.service` notation on earlier forms of CentOS.  Kept reporting that it couldn't find the service file.  I don't have enough knowledge of systemd and init files to know why that may have been the case.

@rootfs I was able to resolve the issue by using the newest versions of CentOS.  I was never deploying using Docker though -- I was using virtualbox.  Looking over your changes, I don't see immediately see anything that would overlap.
